backends/p4tools/testgen/targets/bmv2/p4_refers_to_parser.cpp

```diff
@@ -0,0 +1,152 @@
+                                      const IR::Type* type) {
+    cstring tmp = "";
+    if (table)
+        tmp = currentName + "_key_" + currentKeyName;
+    else
+        tmp = currentName + currentKeyName;
+    auto left = IRUtils::getZombieConst(type, 0, tmp);
+    std::string str = static_cast<std::string>(currentName);
+    std::vector<std::string> elems;
+    std::stringstream ss(str);
+    std::string item;
+    while (std::getline(ss, item, '.')) {
+        elems.push_back(item);
+    }
+    str = "";
+    for (long unsigned int i = 0; i < elems.size() - 1; i++) {
+        str += elems[i] + ".";
+    }
+    tmp = str + destTableName + "_key_" + destKeyName;
+    auto right = IRUtils::getZombieConst(type, 0, tmp);
+    auto expr = new IR::Equ(left, right);
+    std::vector<const IR::Expression*> constraint;
+    constraint.push_back(expr);
+    restrictionsVec.push_back(constraint);
+}
+const IR::Node* RefersToParser::postorder(IR::Annotation* annotation) {
+    const IR::P4Action* prevAction = nullptr;
+    if (annotation->name.name == "refers_to") {
+        if (auto action = findOrigCtxt<IR::P4Action>()) {
+            if (prevAction != action) {
+                actionVector.push_back(action);
+                prevAction = action;
+            }
+        } else if (auto keys = findOrigCtxt<IR::Key>()) {
+            auto table = findOrigCtxt<IR::P4Table>();
+            CHECK_NULL(table);
+            auto key = findOrigCtxt<IR::KeyElement>();
+            CHECK_NULL(key);
+            auto it = find(keys->keyElements.begin(), keys->keyElements.end(), key);
+            if (it != keys->keyElements.end()) {
+                int id = it - keys->keyElements.begin();
+                createRefersToConstraint(key->annotations->annotations, key->expression->type,
+                                         table->controlPlaneName(), id, false,
+                                         key->expression->toString());
+            }
+        }
+    }
+    return annotation;
+}
+
+const IR::P4Action* RefersToParser::findAction(const IR::ActionListElement* input) {
+    for (auto element : actionVector) {
+        if (input->getName().name == element->name.name) {
+            return element;
+        }
+    }
+    return nullptr;
+}
+
+// Builds a variable name from the body of the "refers_to" annotation.
+// The build starts at index 2 because 0 is the table name and 1 is ",".
+cstring buildName(IR::Vector<IR::AnnotationToken> input) {
+    cstring result = "";
+    for (long unsigned int i = 2; i < input.size(); i++) {
+        result += input[i]->text;
+    }
+    return result;
+}
+
+void RefersToParser::createRefersToConstraint(IR::Vector<IR::Annotation> annotations,
+                                              const IR::Type* inputType, cstring controlPlaneName,
+                                              int id, bool isParameter, cstring inputName) {
+    cstring destTableName = "";
```

Comment:
This function can also be simplified.

Reply:
It is already quite simple, of course, complicated by the fact that I tried to unify it for two cases, so as not to duplicate the code. After all, `refers_to ` happens both in the key and in the parameter, because of this, we had to make a function that can support both options. What exactly are you trying to simplify here?

Reply:
For example, you could reduce the nesting by using `continue`. The function does not need to iterate further if the annotation name is not `refers_to`. You can also reset destTableName and destKeyName right at the beginning.